### ISSUE TITLE
Which "pubKeyCredParams" to use?

### ISSUE THREAD

**Issue by dagnelies** [w3c/webauthn#1757]
Hi,

I noticed that during `credentials.create(...)`, if the list does not contain what the authenticator can provide, the authenticator will not be included in the list of authenticators to choose from. For example, if you don't include `"alg":-257`, Windows Hello won't work.

Now, as a relying party this all sounds a bit like unknown mysteries. 

- the specification says "pick your algorithms" from a [huge list](https://www.iana.org/assignments/cose/cose.xhtml#algorithms)!
- no idea which algos the authenticators support
- no idea which algos you really have to support as an RP

In practice, using this list restricts your choice to a subset of authenticators available... if you manage to find out which algo is needed. Also, most RPs are not deeply knowledgeable about which crypto algorithms is better suited or not.

So ...are all common authenticators covered by RS256 and ES256? Or should you as an RP add some more to cover most authenticators? Which ones?

**Comment by MasterKale** [comment]
It's true that you need to explicitly set `-257` to get Hello to work. At Cisco we've managed to support all known authenticators with just `-7` and `-257` in `pubKeyCredParams`. Anecdotal, yes, and unfortunately I don't know if the spec can be so prescriptive...

**Comment by MasterKale** [comment]
And for anyone interested: based on some extensive testing I did a few months back of in-the-wild authenticators, most everything I tested **only** supported `-7` ("ES256"), with the exception of Windows Hello which was only `-257` ("RS256"). Only the YubiKey 5C, 5Ci, and Bio **also** supported Ed25519 (`-8`, "EdDSA").

**Comment by emlun** [comment]
The short answer is that you should set `pubKeyCredParams` to include all algorithms whose verification procedure you support. But that of course leaves the question of what's the minimal set of algorithms you should support.

As @MasterKale points out, ES256 and RS256 is a good start. This is what [`pubKeyCredParams` defaults to](https://www.w3.org/TR/2021/REC-webauthn-2-20210408/#ref-for-dom-publickeycredentialcreationoptions-pubkeycredparams) if you don't specify it. CTAP1 (U2F) authenticators are hard-coded to only support ES256, and most CTAP2 devices support it too.

So I'd recommend at least ES256 and RS256 as a baseline. Beyond that it's up to the RP how much effort to spend on supporting additional algorithms. If you use a library for signature verification, you should probably support everything supported by the library.

**Comment by timcappalli** [comment]
Please use fido-dev@fidoalliance.org (register here: https://groups.google.com/a/fidoalliance.org/g/fido-dev) for implementation discussions. This repo is for the WebAuthn specification itself.

**Comment by cyberphone** [comment]
The idea is to let algorithms evolve including support for quantum safe ditto.

It would though be useful creating a separate "living" algorithm document so that RPs can see what the current options are and which are deprecated.

Have a default doesn't work because then a RP may be get a registration for a "new" algorithm which it isn't yet prepared for.  Therefore specifying the algorithms that the RP understands and accepts remains a viable solution.

**Comment by dagnelies** [comment]
Well, the question is what is better...

1. If the authenticator uses another algorithm than in the `pubKeyCredParams` list, it will simply not appear in the list in the browser pop up. For a user, it is basically like it does not exist and the RP does not know about it.
2. In the other case, if the RP receives an unknown algo, it can simply reply "Sorry, your device is not yet supported" and both the User and the RP know what's going on.

I personnally prefer to be informed and receive explicit errors rather than having the device simply not listed.

Moreover if the empty list is used with `pubKeyCredParams` it will be filled with defaults. Therefore, you might receive new unknown algos anyway. ...and this has already changed in the past rather discretely.

Lastly, I don't think signing algos evolve that quickly and this sounds like YAGNI. For instance, the JWT spec from 2015 still uses the same list of signing algos as today, and there is no change in sight AFAIK.

From a dev POV, it would also make sense if the JWT signing algos could be used, since broad support is already available in all programming languages. It's just nice to have RFCs share a same set of algos (ideally using the same names) instead of different subsets.

**Comment by MasterKale** [comment]
I'm going to turn this into an editorial PR that suggests RP's can benefit from wide authenticator support if they support `-7` `-257`. By having it laid out so explicitly I think it'll help RP's understand and move on to more important questions about WebAuthn use.

**Comment by Firstyear** [comment]
> I noticed that during `credentials.create(...)`, if the list does not contain what the authenticator can provide, the authenticator will not be included in the list of authenticators to choose from. For example, if you don't include `"alg":-257`, Windows Hello won't work.

Worth pointing out that in the windows Hello attestation, there is SHA1 used over the signatures which can be potentially a secuirity risk, so you need to check for the use of RS1 in some internal code paths and reject if found. 

> So ...are all common authenticators covered by RS256 and ES256? Or should you as an RP add some more to cover most authenticators? Which ones?

And ED25519 (EDDSA) is the other one. We have been running a compatibility tester as part of webauthn-rs and "in the wild" we have only seen RS256, ES256 and EDDSA ( '-8' ) .

**Comment by nuno0529** [comment]
For the RP which want to pass FIDO2 server certification, the 3 algos mentioned above are marked as required,  https://fidoalliance.org/specs/fido-v2.0-rd-20180702/fido-server-v2.0-rd-20180702.html#other
I can't say for FIDO if this is an out-of-date document or any strong opinion here, but for sure it's a little useful to your question about compatibility. Besides, I can't see any other spec or document says for that, except https://github.com/w3c/webauthn/blob/main/index.bs#L2772 for user agent's compatibility.  

And for what algos authenticators can support what, a good approach is to check the FIDO MDS and there is a copy of the latest data under https://github.com/opotonniee/fido-mds-explorer/blob/main/mds.js and that is monthly updated from FIDO MDS3.  You can search for the public information about how many other authenticators can support "alg": -8 or other algos. But obviously, MDS3 is optional and it's only for some authenticator vendors which want to have relative certifications and agree to publish the metadata information.

**Comment by MasterKale** [comment]
I'm trying to figure out the best place to put such an editorial blurb. There are currently maybe three places that I see as candidates:

1. Sample code in the registration examples: https://www.w3.org/TR/webauthn-2/#sctn-sample-registration
2. Step 10 in `createCredential()` steps: https://www.w3.org/TR/webauthn-2/#sctn-createCredential
3. This section on Easily Accessing Credential Data: https://www.w3.org/TR/webauthn-2/#sctn-public-key-easy
4. The section in `PublicKeyCredentialCreationOptions` about `pubKeyCredParams`: https://www.w3.org/TR/webauthn-2/#dom-publickeycredentialcreationoptions-pubkeycredparams

I'd like to put this addition in #4 as it feels like it has the best chance of an RP dev seeing it as they're trying to understand all the options. That said, it'd mean creating a _fourth_ place to need to keep up-to-date as future algorithms get introduced. If putting an editorial suggestion to use `-256` and `-257` (and maybe `-8` now) here has the greatest chance of being the most clear recommendation on which algorithms an RP should put into `pubKeyCredParams` (or suggesting _not_ specifying a value for `pubKeyCredParams` so the defaults in #2 are used), then maybe I include further changes to these other sections to reference #4. Thoughts?

BTW #2 is missing a reference to `-8` that exists in #3, is that something we should reconcile? And perhaps we should consider updating #2 (in a separate issue) to prioritize Ed25519 keypairs (by putting `-8` first) now that it's in the wild 🤔

**Comment by Firstyear** [comment]
@MasterKale Or you add a new section and link to it from those locations so that it's only one place? Saying that it's a bit of a weird situation because the webauthn spec is the "only dev guide" today, when really this is an implementation detail and requires more than just "use these algos" because there is a *lot* going on.

**Comment by MasterKale** [comment]
> ...this is an implementation detail and requires more than just "use these algos" because there is a lot going on.

It doesn't help that the spec is prescriptive about specific algorithms (`-256`, `-257`, and `-8`), but leaves the door open for support of other algorithms (that RP's would have to somehow track independent of the spec.)

**Comment by emlun** [comment]
@MasterKale by `-256`, did you mean `-7` (P-256)?

I think (4) [`pubKeyCredParams`](https://w3c.github.io/webauthn/#dom-publickeycredentialcreationoptions-pubkeycredparams) is the most appropriate place to put this guidance. I agree that's where an RP dev is most likely to see it. (1) should ideally agree with whatever we put in (4), but shouldn't be the primary authoritative source. I don't think this is necessarily related to the algs list in (3). (2) and (4) should probably be the same, but I'm not sure if it's appropriate to have them reference each other. But if so, (2) should probably reference (4) rather than vice versa.

It would be nice to recommend `[]` to fall back to the default, but it's not backward-compatible with browser implementations still on L1, so I'm not sure it's a good idea. (...although I suppose you _could_ feature-detect on e.g. `window.AuthenticatorAttestationResponse.prototype.getPublicKey === undefined` to distinguish L1 clients. But that gets pretty involved - and in the end you'd still need to choose a `pubKeyCredParams` value for L1, so you might as well always use it.)

**Comment by WordlessEcho** [comment]
For developer: **tldr** use -7 (ES256)

I do some research of Windows Hello (22H2), Apple Passkey (iOS 16.2 (20B82)), and Android Passkey (Chrome Canary 110.0.5428.0, Android 11 RSR1.201211.001, GMS 22.44.17 (150800-488410709)).
**On Windows** (`authenticatorAttachment: 'platform'`): -257 (RS256) and -7 (ES256) is supported. If you specify unsupported algorithm, you will get `DOMException: The operation either timed out or was not allowed. See: https://www.w3.org/TR/webauthn-2/#sctn-privacy-considerations-client.` on Chrome and `DOMException: The operation failed for an unknown transient reason` on Firefox.
**On iOS**: -7 (ES256) is supported. Passkey will prompt "can not finish this operation" after you finish biometrics verification.
**On Android**:  -7 (ES256) is supported. ``DOMException: None of the algorithms specified in `pubKeyCredParams` are supported by this device.``

For hardware key (tested by Yubikey 5C, FIDO 2 PIN is set, firmware 5.4.3):
**On Windows**: If you leave `authenticatorAttachment` default with unsupported algorithm, Windows Hello will directly ask for hardware key. If unsupported algorithm for key is specified, an error message will be flashed once quickly then return to ask PIN for key (loop).
**On iOS** (via NFC): If unsupported algorithm for key is specified, you will stuck in "insert your key". No indicates or prompt.
**On Android**: You won't see any interface if you specify unsupported algorithm for Android. Even if `authenticatorAttachment` is set to `cross-platform`. So I can't continue to test because I don't have hardware key that doesn't support -7 (ES256).

**Comment by MasterKale** [comment]
> For developer: **tldr** use -7 (ES256)

The three algorithms I've thought lately we can safely suggest within the spec are, in this order:

- `-8` (Ed25519)
- `-7` (ES256)
- `-257` (RS256)

I'd put `-8` at the top of the list to encourage adoption of Ed25519, since authenticators that can support it will choose to use it if it's first on the list. After that `-7` and `-257` are the two that practically cover almost 100% of existing authenticators.

I wouldn't drop `-257` just yet, even if signs are that platform authenticators used via Windows Hello will start using ES256, because of how long older versions of Windows can stick around.

Thanks for bumping this up, it reminded me that I need to get started on this 😅

**Comment by Firstyear** [comment]
We have issues with ed25519 in webauthn-rs just because it's not properly supported in the ecosystem cryptographically, but that's "our problem". 

RS256 is okay, but the risk is sometimes they'll be signed with RS1 (yes, sha 1) so implementors need to be careful of that.

**Comment by dagnelies** [comment]
> For developer: **tldr** use -7 (ES256)

Well, depending on your Windows version, you might fall on your face. Until recently only RS256 was supported for Windows Hello. Dunno when exactly this changed.

There is also quite some discrepancy between the "real world" and the implementations of the "big three" and the specs here, and both evolve at different paces. It's already good that the spec trimmed down the recommended algos to three.

**Comment by petrdvorak** [comment]
... I retroactively wonder what is the benefit of recommending multiple algorithms... More work, less compatibility, slower adoption... Can you see any hard benefits over just recommending: "Use ES256 at a minimum" (which is universally supported in all new systems)? Mixing RSA with EC in one codebase on a server just adds complexity, while almost nobody ends up using RSA these days so the complexity (as I see it) does not bring much value.

**Comment by petrdvorak** [comment]
To better explain my perspective... We are finalizing a new FIDO2 product and we will need to end up building abstraction for legacy RS256 and still a bit "exotic" Ed25519 just so nobody ends up using it when we are done. The recommendation just puts additional dead weight in our software.

**Comment by MasterKale** [comment]
Windows Hello still uses RS256, and of course it won't be an overnight transition to anything else even if Microsoft's passkeys support flips Windows over to using ES256 exclusively. RP's have to build to what's in the market, and so I think it's still important to help RP's support use of RSA with WebAuthn even if it's seen as "legacy support".

**Comment by petrdvorak** [comment]
@MasterKale Understood, and thank you for the prompt response. 👍  I am thinking the recommendation could maybe be worded as:

>         [=[RPS]=] MUST support the following {{COSEAlgorithmIdentifier}} value:
> 
>         * -7 (ES256)
> 
>        [=[RPS]=] that wish to support a wide range of [=authenticators=], including some legacy ones,
>        SHOULD also include at least the following {{COSEAlgorithmIdentifier}} values:
> 
>         * -8 (Ed25519)
>         * -257 (RS256)
> 
>         Additional signature algorithms can be included as needed.

This approach will both:

- Provide minimum mandatory supported algorithm.
- Recommend algorithms to achieve broader support.

ES256 is already widely adopted, not only in WebAuthn but also when working with JWTs. I think making this the default could speed up WebAuthn adoption (as even small website makers will be able to provide minimalistic implementations). I will first sleep on it to see if this sounds like a good idea in the morning 🙂  and if it does, I will try to draft a PR - in the worst case, it gets rejected with some comments, which is fine.

**Comment by emlun** [comment]
I think that in practice, the majority of RPs will most likely use an external library to implement WebAuthn signature verification. In that case they usually shouldn't have to do anything extra to support the additional algorithms, all of that work is on the library developer.

The main exception I can see is platforms like Java where you need to provide JCA providers for each algorithm, but even there the common ones are built in on recent JRE versions.

**Comment by petrdvorak** [comment]
@emlun Agreed. We are makers of such a back-end system for banks, considering adding WebAuthn support and seeing close to zero benefits in implementing RS256 and Ed25519 because these will be rarely used. Plus, we have an idea on how to build WebAuthn back-end in a more creative way to simplify enrollment to service providers, and we are reluctant to support RS256 there because of the "old Windows problem."

Still need more sleep actually to make any move there. Maybe we are too "pragmatic" and I sure understand the "broader industry view" to formally support more algorithms despite not being really used.

**Comment by MasterKale** [comment]
> We are makers of such a back-end system for banks, considering adding WebAuthn support and seeing close to zero benefits in implementing RS256 and Ed25519 because these will be rarely used. Plus, we have an idea on how to build WebAuthn back-end in a more creative way to simplify enrollment to service providers, and we are reluctant to support RS256 there because of the "old Windows problem."

This is starting to get into "product requirement" territory. If you decide you don't want to support RS256 or Ed25519 then that's your decision as the product owner. Your customers will ultimately decide whether that's viable long-term. 

However "supporting as many versions of Windows as possible" is likely to be a requirement for the majority of other projects. Therefore I'd suggest we keep the current guidance as-is because the spec should aim for maximum support of the API.

**Comment by petrdvorak** [comment]
@MasterKale Yes, I get that... But then why not support all other algorithms on the same level? This would broaden the range of supported authenticators, right? I think this is one of the situations of "if we do not specify default choice, people will just make choices and use everything" (which - again - might be right). I don't want to push this too hard - you are doing an excellent job with the activity and we are excited about WebAuthn. But looking at legacy Windows authenticators (people with old Windows can still use Passkeys or USB authenticators) is exactly the type of anchor that will eventually hold the activity back.

**Comment by dagnelies** [comment]
@petrdvorak we're not looking at "old Windows legacy systems", it's about the "latest Windows 11" 😉

**Comment by petrdvorak** [comment]
@dagnelies I will do some testing (... Mac user here 🐑 ) but Hello should support ES256 too.

**Comment by MasterKale** [comment]
I think I remember being surprised by an ES256 response from Windows 11 at some point, but with Windows 10 and slower-updating (enterprise) installs of Windows 11 still out in the wild, I don't think RS256 is going away any time soon.

> I think this is one of the situations of "if we do not specify default choice, people will just make choices and use everything"

But the L3 spec now does suggest a default choice:

https://w3c.github.io/webauthn/#dom-publickeycredentialcreationoptions-pubkeycredparams

This is, admittedly, based on observation of actual algorithms being returned out in the real world (`-7` and `-257`), with a sprinkling of "this is great to use if it's available" (`-8`.) If at some point an algorithm stops being used, or a newer, better algorithm gains traction in the authenticator space, then we should absolutely update this list.

**Comment by petrdvorak** [comment]
@MasterKale Agreed. Chrome actually complains when JavaScript does not request both `-7` and `-257`, so as far as Google is concerned, these two are the minimum.

<img width="1512" alt="Screenshot 2023-05-03 at 18 31 40" src="https://user-images.githubusercontent.com/516597/235980868-fb329e24-83e4-4d01-8445-f9a3272f3d04.png">

The choice of `-8` is pleasing to the "cryptographer me" (the algorithm is indeed "interesting"), but I am a bit puzzled about why this is the first choice on the list as this is not really reflected by any dominating position of the algorithm on the market...

**Comment by destinysz** [comment]
> 对于任何感兴趣的人：根据几个月前我对野外身份验证器进行的一些广泛测试，我测试的大多数内容都**仅**支持`-7`（“ES256”），但 Windows Hello 除外，它仅支持`-257`（“RS256”） 。只有 YubiKey 5C、5Ci 和 Bio**还**支持 Ed25519（`-8`“EdDSA”）。

Has yubikey security key nfc been tested to support ed25519?

**Comment by antonymott** [comment]
@MasterKale, petrdvorak and other contributors: thanks for this discussion which I find very useful, and even though closed now, perhaps useful to add April 2024 update on my experience with algorithm order.

I'm doing webauthn from scratch with a library mentality, and expect most of my customers to be early-adopters or mobile users with standard platform biometrics (touch or faceid). I thought about not supporting ancient windows Hello users, but I also had those chrome warnings when I left out the legacy -257.

I followed Matthew's suggestion and reorganized my algs to put -8 first, to encourage the more current algorithms to bubble up...sure enough latest Chrome used -8, and ignored -7 which was second in my alg list. I did keep last the -257 alg.

I store the public-key algorithm, so over the next year or so I can measure what percentage if any of my customers will create keys with -257. If anyone is interested shout out to me and I'll post the metrics here or somewhere more useful.

Exciting technology: the way forward!

**Comment by coolaj86** [comment]
> I'd put -8 at the top of the list to encourage adoption of Ed25519, since authenticators that can support it will choose to use it if it's first on the list.

Can someone perhaps get a message over to the WebCrypto people to make a stronger recommendation for Ed25519 support in the browser itself?

I'm looking at 1mb of libsodium in this demo, and that's a big deterrent to adoption for me.
Demo: https://mylofi.github.io/webauthn-local-client/

Signing the JWTs and whatever else without the native browser support makes for a lot of bloat.

**Comment by antonymott** [comment]
@MasterKale and to anyone else interested, here's an in-the-wild stats update to my April 2 comment that indicates more adoption of EdDSA. Thanks partly to MasterKale's useful youTube vids and various github comments, my authenticator listed 3 algs in the order [-8,-7,-257]. After 5 months of in-the-wild usage with approx 900 users, roughly half authenticated with ES256 (-7), roughly half with EdDSA (-8) and less than 1% used legacy -257.

**Comment by MasterKale** [comment]
@antonymott Thanks for reporting back! That's very encouraging to see how commonly Ed25519 is being supported! Have you tried to understand how many of those are security keys vs platform authenticators? If I had to guess all the `-8` are security keys 🤔

**Comment by tylerccarson** [comment]
Given all the investigation of algorithm support on the various authenticators hoping someone has some insight-- Windows 11 added ECC algorithm support [last year](https://learn.microsoft.com/en-us/windows/whats-new/whats-new-windows-11-version-22h2#webauthn-apis-support-ecc) (ES256). What's surprising is that there are some Windows 11 installations, even when up to date, that still only allow creating passkeys with RS256. Is there any reasonable explanation for this?

**Comment by petrdvorak** [comment]
On another note... we can now use Dilithium signatures (ML-DSA), so I would advocate for adding those.

https://www.wultra.com/blog/passkeys-and-fido2-quietly-became-quantum-safe-heres-what-changed
